src/map/besieged_data.cpp

```diff
@@ -0,0 +1,89 @@
+#include "besieged_data.h"
+
+#include "common/settings.h"
+
+BesiegedData::BesiegedData(std::unique_ptr<SqlConnection>& sql)
+: strongholdInfos(std::vector<stronghold_info_t>(4))
+{
+    load(sql);
+}
+
+void BesiegedData::load(std::unique_ptr<SqlConnection>& sql)
+{
+    const char* Query = "SELECT stronghold_id, orders, stronghold_level, forces, mirrors, prisoners, mirror_destroyed, owns_astral_candescence FROM besieged_strongholds";
+    int32       ret   = sql->Query(Query);
+
+    if (ret != SQL_ERROR && sql->NumRows() != 0)
+    {
+        while (sql->NextRow() == SQL_SUCCESS)
+        {
+            uint8 strongholdId = sql->GetUIntData(0);
+            if (strongholdId > 3)
+            {
+                ShowError("Invalid stronghold id %d", strongholdId);
+                throw std::runtime_error("Besieged stronghold id out of range");
+            }
+
+            stronghold_info_t strongholdInfo;
+            strongholdInfo.orders                = (BEASTMEN_BESIEGED_ORDERS)sql->GetUIntData(1);
+            strongholdInfo.stronghold_level      = (STRONGHOLD_LEVEL)sql->GetUIntData(2);
+            strongholdInfo.forces                = sql->GetUIntData(3);
+            strongholdInfo.mirrors               = sql->GetUIntData(4);
+            strongholdInfo.prisoners             = sql->GetUIntData(5);
+            strongholdInfo.mirrorDestroyed       = sql->GetUIntData(6);
+            strongholdInfo.ownsAstralCandescence = sql->GetUIntData(7);
+
+            strongholdInfos[sql->GetUIntData(0)] = strongholdInfo;
+        }
+    }
+
+    if (settings::get<bool>("logging.DEBUG_BESIEGED"))
+    {
+        ShowInfo("Loading besieged Stronghold Data:");
+        for (uint8 strongholdId = 0; strongholdId < strongholdInfos.size(); strongholdId++)
+        {
+            stronghold_info_t strongholdInfo = strongholdInfos[strongholdId];
+            ShowInfo("Stronghold %d: Orders %d, Level %d, Forces %d, Mirrors %d, Prisoners %d, Mirror Destroyed %d, Owns Astral Candescence %d",
+                     strongholdId, strongholdInfo.orders, strongholdInfo.stronghold_level, strongholdInfo.forces, strongholdInfo.mirrors, strongholdInfo.prisoners, strongholdInfo.mirrorDestroyed, strongholdInfo.ownsAstralCandescence);
+        }
+    }
+}
+
+BESIEGED_STRONGHOLD BesiegedData::getAstralCandescenceOwner()
+{
+    for (uint8 strongholdId = 0; strongholdId < strongholdInfos.size(); strongholdId++)
+    {
+        if (strongholdInfos[strongholdId].ownsAstralCandescence)
+        {
+            return (BESIEGED_STRONGHOLD)strongholdId;
+        }
+    }
+
+    ShowError("No stronghold owns the Astral Candescence");
+    return BESIEGED_STRONGHOLD::ALZAHBI;
+}
+
+ALZHABI_BESIEGED_ORDERS BesiegedData::getAlZhabiOrders()
+{
+    return (ALZHABI_BESIEGED_ORDERS)strongholdInfos[BESIEGED_STRONGHOLD::ALZAHBI].orders;
+}
+
+stronghold_info_t BesiegedData::getStrongholdInfo(BESIEGED_STRONGHOLD stronghold)
+{
+    if (stronghold > strongholdInfos.size())
```

Comment:
Does this c rash if you ask for `strongholdInfos.size()`, zero-indexing and all that

Reply:
what do you mean by asking for strongholdInfos.size()?

Reply:
stronghold passed in here is an enum that can only be 0-3, strongholdInfos.size() is always going to be 4 - this check is only required if strongholdInfos could possibly be different?

Reply:
oh I see. That's true. I could remove that check.

Reply:
come to think about it, the reason I added this check is because if the db is incorrectly setup (more than 4 strongholds, then we would crash here without knowing why.